pdns/dnsrecords.cc

```diff
@@ -489,6 +490,195 @@ string EUI64RecordContent::getZoneRepresentation(bool noDot) const
 
 /* EUI64 end */
 
+/* APL start */
+/* https://tools.ietf.org/html/rfc3123 */
+void APLRecordContent::report(void)
+{
+  regist(1, QType::APL, &make, &make, "APL");
+}
+// Parse incoming packets (e.g. nsupdate)
+std::shared_ptr<DNSRecordContent> APLRecordContent::make(const DNSRecord &dr, PacketReader& pr) {
+  uint8_t temp;
+
+  if(dr.d_clen < 5 or dr.d_clen > 20)
+    throw MOADNSException("Wrong size for APL record");
+
+  auto ret=std::make_shared<APLRecordContent>();
+  pr.xfr16BitInt(ret->d_family);
+  pr.xfr8BitInt(ret->d_prefix);
+  pr.xfr8BitInt(temp);
+  ret->d_n = (temp & 128) >> 7;
+  ret->d_afdlength = temp & 127;
+
+  if (ret->d_family == APL_FAMILY_IPV4) { // IPv4
+    if (ret->d_afdlength > 4) {
+      throw MOADNSException("Invalid IP length for IPv4 APL");
+    }
+    bzero(ret->d_ip4, sizeof(ret->d_ip4));
+    for (int i=0; i < 4; i++) {
+      if (i < ret->d_afdlength)
+        pr.xfr8BitInt(ret->d_ip4[i]);
+    }
+  } else if (ret->d_family == APL_FAMILY_IPV6) {
+    if (ret->d_afdlength > 16) {
+      throw MOADNSException("Invalid IP length for IPv6 APL");
+    }
+    bzero(ret->d_ip6, sizeof(ret->d_ip6));
+    for (int i=0; i< 16; i++) {
```

Comment:
why not just `i < min(16, ret->d_afdlength)` ?

Reply:
Done in Jan 2nd commits